### PR TITLE
feat: for taker fee, search for route with dym to swap and burn

### PR DESCRIPTION
**Pull request by mtsitrin**

### CODE REVIEW

File: x/gamm/keeper/taker_fee.go [hunk 1]
```diff
@@ -11,104 +14,91 @@ const (
...
-		}
-	}
-
-	//if not found, return empty as we can't swap to the defaultTakerFeeDenom
-	if idx == 0 || !found {
-		return newRoutes
+// possibilites for takerFeeCoin:
+// 1. DYM -> taker fee is already in DYM
+// 2. any other denom -> DYM (taker fee will be swapped and burned)
+// 2. Base denom -> find pool with DYM
```
Comment (omritoptix) [inline x/gamm/keeper/taker_fee.go]: ordering - you have twice `2.`

File: x/gamm/keeper/taker_fee.go [hunk 2]
```diff
@@ -11,104 +14,91 @@ const (
...
-	if err != nil {
-		return err
-	}
-	return nil
 }
 
-func (k Keeper) chargeTakerFeeSwapAmountIn(ctx sdk.Context, takerFeeCoin sdk.Coin, sender sdk.AccAddress, routes []poolmanagertypes.SwapAmountInRoute) error {
-	if takerFeeCoin.Denom == defaultTakerFeeDenom {
-		return k.burnTakerFee(ctx, takerFeeCoin, sender)
+func (k Keeper) swapAndBurn(ctx sdk.Context, sender sdk.AccAddress, routes []poolmanagertypes.SwapAmountInRoute, tokenIn sdk.Coin) error {
```
Comment (omritoptix) [inline x/gamm/keeper/taker_fee.go:56]: could be useful to add descirption about what this function is supposed to do

File: x/gamm/keeper/taker_fee.go [hunk 3]
```diff
@@ -11,104 +14,91 @@ const (
...
 	}
 
-	return newRoutes
+	return k.swapAndBurn(ctx, sender, []poolmanagertypes.SwapAmountInRoute{firstPool}, takerFeeCoin)
 }
 
-func (k Keeper) chargeTakerFeeSwapAmountOut(ctx sdk.Context, takerFeeCoin sdk.Coin, sender sdk.AccAddress, outRoutes []poolmanagertypes.SwapAmountOutRoute, denomOut string) error {
+// possibilites for takerFeeCoin:
+// 1. DYM -> taker fee is already in DYM
+// 2. any other denom -> DYM (taker fee will be swapped and burned)
```
Comment (omritoptix) [inline x/gamm/keeper/taker_fee.go:41]: ordering - you have twice 2.

File: x/gamm/keeper/taker_fee.go [hunk 4]
```diff
@@ -11,104 +14,91 @@ const (
...
 
-	for i, route := range routes[:idx] {
-		newRoutes = append(newRoutes, poolmanagertypes.SwapAmountInRoute{
-			PoolId:        route.PoolId,
-			TokenOutDenom: routes[i+1].TokenInDenom,
-		})
+	//transcode the outRoutes to inRoutes
+	firstPool := poolmanagertypes.SwapAmountInRoute{}
+	firstPool.PoolId = outRoutes[0].PoolId
+	if len(outRoutes) > 1 {
```
Comment (omritoptix) [inline x/gamm/keeper/taker_fee.go:30]: is there a chance that the len(routes) > 2? what gurantees that the frontend won't send longer routes?

File: x/gamm/keeper/taker_fee.go [hunk 5]
```diff
@@ -143,3 +133,29 @@ func (k Keeper) burnTakerFee(ctx sdk.Context, takerFeeCoin sdk.Coin, sender sdk.
 
 	return nil
 }
+
+// find route from denom to defaultTakerFeeDenom by iterating all pools
+func (k Keeper) findPoolWithDYM(ctx sdk.Context, fromDenom string) (poolmanagertypes.SwapAmountInRoute, error) {
```
Comment (omritoptix) [inline x/gamm/keeper/taker_fee.go]: prob better name for func would be findPoolWithTakerFeeDenom

**Review by omritoptix — COMMENTED** [review]
is the new extra potential  fee from swapping the taker fee to DYM is calculated in the estimation functions for calculating the end amount for the users ? 
I believe they are called 'CalcOutAmtGivenIn` and vice versa. 

didn't see it here.

File: x/gamm/keeper/taker_fee.go (hunk 4 above)
Comment (mtsitrin) [reply x/gamm/keeper/taker_fee.go:30]: we still need only the first pool.
one of this assets will be base denom
if not dym (e.g usdc), we'll find a pool for usdc->dym

**Review by omritoptix — APPROVED** [review]

**Pull request merged**